url/urlsearchparams.go

```diff
@@ -0,0 +1,368 @@
+package url
+
+import (
+	"fmt"
+	"net/url"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/dop251/goja"
+)
+
+var (
+	reflectTypeString = reflect.TypeOf("")
+	reflectTypeObject = reflect.TypeOf(map[string]interface{}{})
+	reflectTypeArray  = reflect.TypeOf([]interface{}{})
+	reflectTypeMap    = reflect.TypeOf([][2]interface{}{})
+)
+
+func newInvalidTypleError(r *goja.Runtime) *goja.Object {
+	return newError(r, "ERR_MISSING_ARGS", "Each query pair must be an iterable [name, value] tuple")
+}
+
+func newMissingArgsError(r *goja.Runtime, msg string) *goja.Object {
+	return newError(r, "ERR_MISSING_ARGS", msg)
+}
+
+func newInvalidArgsError(r *goja.Runtime) *goja.Object {
+	return newError(r, "ERR_INVALID_ARG_TYPE", `The "callback" argument must be of type function. Received undefined`)
+}
+
+func newError(r *goja.Runtime, code string, msg string) *goja.Object {
+	o := r.NewTypeError("[" + code + "]: " + msg)
+	o.Set("code", r.ToValue(code))
+	return o
+}
+
+// Currently not supporting the following:
+//
+//   - ctor(iterable): Using function generators
+func createURLSearchParamsConstructor(r *goja.Runtime) goja.Value {
+	f := r.ToValue(func(call goja.ConstructorCall) *goja.Object {
+		u, _ := url.Parse("")
+		if len(call.Arguments) > 0 {
+			p := call.Arguments[0]
+			e := p.Export()
+			switch p.ExportType() {
```

Comment:
ExportType meant as an optimisation that allows to determine the exported type without doing actual export (which can be quite costly, consider an Object with lots of fields).

Therefore, doing unconditional Export() and then calling ExportType() makes no sense: either do Export() and type switch (and risk wasting resources if the value is not of the right type) or call ExportType() and then only do Export() if it's the right type.